### PR TITLE
port tests module to cats-effect 3

### PR DESCRIPTION
**Pull request by yanns**
https://github.com/http4s/http4s/issues/3835

### CODE REVIEW

File: tests/src/test/scala/org/http4s/EntityCodecSpec.scala [hunk 1]
```diff
@@ -32,3 +32,4 @@ class EntityCodecSpec extends Http4sSpec {
 
   checkAll("EntityCodec[IO, Unit]", EntityCodecTests[IO, Unit].entityCodec)
 }
+ */
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/EntityCodecSpec.scala:35]: @cquiroz I've lost track... do any of these new libraries floating around solve this problem?

File: tests/src/test/scala/org/http4s/EntityDecoderSpec.scala [hunk 2]
```diff
@@ -301,7 +301,7 @@ class EntityDecoderSpec extends Http4sSpec with Http4sLegacyMatchersIO with Pend
 
     "invoke the function with  the right on a success" in {
       val happyDecoder: EntityDecoder[IO, String] =
-        EntityDecoder.decodeBy(MediaRange.`*/*`)(_ => DecodeResult.success(IO.pure("hooray")))
+        EntityDecoder.decodeBy(MediaRange.`**`)(_ => DecodeResult.success(IO.pure("hooray")))
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/EntityDecoderSpec.scala:304]: Is this change to make it compile inside the block comment?  You can put another `*/` at the end, which looks imbalanced, but works.

Or at least you could in Scala 2.9 or whatever when I first had this problem. :laughing:

File: tests/src/test/scala/org/http4s/EntityCodecSpec.scala (hunk 1 above)
Comment (yanns) [reply tests/src/test/scala/org/http4s/EntityCodecSpec.scala]: By adding an instance for `Eq[F[_]]`, this test can run: https://github.com/http4s/http4s/pull/4007/commits/c19cdd2093cd313ef9fb1d2e308af337a7dfba74

Do you think it's a suitable solution?

Comment (rossabaker) [reply tests/src/test/scala/org/http4s/EntityCodecSpec.scala]: Brief discussion on [cats-effect-dev Gitter](https://gitter.im/typelevel/cats-effect-dev?at=5fd4f24628a57c581b0bfb66).

I think this is fine for now, but I think in the long run Typelevel needs a discipline-effect built around scalacheck-effect's `PropF`.

**Comment by yanns** [comment]
The remaining tests to be ported are blocked by the following error:
```
==> X org.http4s.MessageSuite.initializationError  0.034s java.lang.NoSuchMethodError: 'java.lang.Object io.chrisdavenport.vault.Key$.newKey(cats.effect.kernel.Sync)'
```

https://github.com/ChristopherDavenport/vault/issues/159

File: core/src/main/scala/org/http4s/StaticFile.scala [hunk 3]
```diff
@@ -154,8 +154,8 @@ object StaticFile {
       etagCalc <- etagCalculator(f).map(et => ETag(et))
       res <- Files[F].isFile(f.toPath()).flatMap[Option[Response[F]]] { isFile =>
         if (isFile) {
-
-          if (start >= 0 && end >= start && buffsize > 0) {
+          val requireCondition = start >= 0 && end >= start && buffsize > 0
+          if (!requireCondition) {
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/StaticFile.scala:158]: Was this broken before?  This looks correct, but inverted from the previous?

**Review by rossabaker — COMMENTED** [review]
Thanks! Glad to see this one about ready to land.

Should we be using `Sync.blocking` where we used to use a `testBlocker`, or is it okay to drop those?

File: core/src/main/scala/org/http4s/StaticFile.scala (hunk 3 above)
Comment (yanns) [reply core/src/main/scala/org/http4s/StaticFile.scala:158]: It was inverted during the ce3 migration: https://github.com/http4s/http4s/commit/31b54aa0ce46a8450cf94eb1d0ddea7aeeacf6a8#diff-ffd60bd1163f5dd37a78811adfe0b03b0462f98d4c58041c46d11089232fc336L143-L145

From:
```
          require(
            start >= 0 && end >= start && buffsize > 0,
            s"start: $start, end: $end, buffsize: $buffsize")
```
to
```
          if (start >= 0 && end >= start && buffsize > 0) {
            F.raiseError[Option[Response[F]]](
              new IllegalArgumentException(
                s"requirement failed: start: $start, end: $end, buffsize: $buffsize"))
```
which is the opposite.

**Comment by yanns** [comment]
> Should we be using `Sync.blocking` where we used to use a `testBlocker`, or is it okay to drop those?

I understood that we could drop those, but I'm not entirely sure. I'm asking on the ce channel.

**Comment by RaasAhsan** [comment]
I'm running into this on the #3857 PR issue as well. Short answer: `Sync.blocking` is intended to be the successor to `Blocker`, but it isn't a complete replacement. `Blocker` lets you evaluate arbitrary effects on another pool (in the sense of `evalOn`), however, `Sync.blocking` only lets you evaluate a thunk `=> A` on the pool. It seems like http4s tends to run large, composite effects on the blocking pool, but for at least the cases I've seen, the truly blocking operations do happen on the blocking pool (via fs2).

So, I think we should do the following:
1. whether `Sync.blocking` is an appropriate alternative to where `Blocker` was used before
2. do we actually need to run large effects on the blocking pool or just simple thunks? the tradeoff here is we incur a lot of overhead from switching between pools if we need to evaluate and compose a bunch of blocking actions
3. If 1 and 2 don't pan out, we'll probably need to reintroduce something like `Blocker` via a separate `ExecutionContext`. maybe worth taking back upstream

**Comment by djspiewak** [comment]
> do we actually need to run large effects on the blocking pool or just simple thunks? the tradeoff here is we incur a lot of overhead from switching between pools if we need to evaluate and compose a bunch of blocking actions

This is definitely something I'm interested in understanding better. `blocking` was very prescriptively designed with the assertion that this is not a valid use-case. If http4s disproves that assertion then we need a different API.

**Review by rossabaker — APPROVED** [review]
No, I think these are all legitimate.  We were threading all these blockers through to `fs2.io`, and it no longer requires them because it is using either `Sync.blocking` or NIO.  The objection was just my ignorance of CE3 and FS2-3.

**Pull request merged**